adcommon/yldap.py

```diff
@@ -175,3 +180,126 @@ def ldap_delete(self, *args):
             ycpbuiltins.y2error(traceback.format_exc())
             ycpbuiltins.y2error('ldap.delete_s: %s\n' % self.__ldap_exc_msg(e))
 
+    def __find_inferior_classes(self, name):
+        dn = 'CN=Schema,CN=Configuration,%s' % self.realm_dn
+        search = '(|(possSuperiors=%s)(systemPossSuperiors=%s))' % (name, name)
+        return [item[-1]['lDAPDisplayName'][-1] for item in self.ldap_search_s(dn, SCOPE_SUBTREE, search, ['lDAPDisplayName'])]
+
+    def __load_schema(self):
+        dn = self.l.search_subschemasubentry_s()
+        results = self.l.read_subschemasubentry_s(dn)
+
+        self.schema['attributeTypes'] = {}
+        self.schema['constructedAttributes'] = self.__constructed_attributes()
+        for attributeType in results['attributeTypes']:
+            m = re.match(b'\(\s+(?P<id>[0-9\.]+)\s+NAME\s+\'(?P<name>[\-\w]+)\'\s+(SYNTAX\s+\'(?P<syntax>[0-9\.]+)\'\s+)?(?P<info>.*)\)', attributeType)
+            if m:
+                name = m.group('name')
+                self.schema['attributeTypes'][name] = {}
+                self.schema['attributeTypes'][name]['id'] = m.group('id')
+                self.schema['attributeTypes'][name]['syntax'] = m.group('syntax')
+                self.schema['attributeTypes'][name]['multi-valued'] = b'SINGLE-VALUE' not in m.group('info')
+                self.schema['attributeTypes'][name]['collective'] = b'COLLECTIVE' in m.group('info')
+                self.schema['attributeTypes'][name]['user-modifiable'] = b'NO-USER-MODIFICATION' not in m.group('info')
+                if b'USAGE' in m.group('info'):
+                    usage = re.findall(b'.*\s+USAGE\s+(\w+)', m.group('info'))
+                    self.schema['attributeTypes'][name]['usage'] = usage[-1] if usage else 'userApplications'
```

Comment:
could be wrong but I thought ldap attributes should be utf8 encoded bytes not string so shouldn't this line be something like

                    self.schema['attributeTypes'][name]['usage'] = usage[-1] if usage else b'userApplications'
and similar other instances in the code (same of course for contents of usage if applicable)